### PR TITLE
Add API to enable/disable HDCP support from HWCService API.

### PR DESCRIPTION
**Pull request by kalyankondapally**
Jira: None.
Test: We are able to toggle HDCP state via the API tested by
      hwcservice_test.
Signed-off-by: Kalyan Kondapally <kalyan.kondapally@intel.com>

### CODE REVIEW

File: os/android/libhwcservice/hwcserviceapi.h [hunk 1]
```diff
@@ -169,6 +152,35 @@ status_t HwcService_DisplayMode_SetMode(HWCSHANDLE hwcs, uint32_t display,
 
 // VideoControl
 
+// The control enables the usage of HDCP for all planes supporting this feature
+// on display. Some displays can support latest HDCP specification and also
+// have ability to fallback to older specifications i.e. HDCP 2.2 and 1.4
+// in case latest speicification cannot be supported for some reason. Fallback
```
Comment (MicheleL-Intel) [inline os/android/libhwcservice/hwcserviceapi.h]: fix typo 'specification'

File: os/android/libhwcservice/hwcserviceapi.h [hunk 2]
```diff
@@ -169,6 +152,35 @@ status_t HwcService_DisplayMode_SetMode(HWCSHANDLE hwcs, uint32_t display,
...
+status_t HwcService_Video_EnableHDCPSession_ForDisplay(HWCSHANDLE hwcs,
+                                                       uint32_t display,
+                                                       bool fallback);
+
+// The control enables the usage of HDCP for all planes supporting this feature
+// on all connected displays. Some displays can support latest HDCP
+// specification
+// and also have ability to fallback to older specifications i.e. HDCP 2.2 and
+// 1.4
+// in case latest speicification cannot be supported for some reason. Fallback
```
Comment (MicheleL-Intel) [inline os/android/libhwcservice/hwcserviceapi.h]: Typo 'specification' and remove extraneous carriage returns or line feeds.

File: common/core/gpudevice.cpp [hunk 3]
```diff
@@ -524,6 +524,43 @@ void GpuDevice::HandleHWCSettings() {
...
 
+void GpuDevice::EnableHDCPSessionForDisplay(uint32_t display,
+                                            bool /*fallback*/) {
+  size_t size = total_displays_.size();
+  if (size <= display) {
+    ETRACE("Tried to enable HDCP for invalid display %d \n", display);
+    return;
+  }
+
+  NativeDisplay *native_display = total_displays_.at(display);
```
[the remark this answers is not in the document]
Comment (kalyankondapally) [reply common/core/gpudevice.cpp]: This should be handled by above checks where we ensure index is not greater than the size of vector

File: os/android/hwcservice.cpp [hunk 4]
```diff
@@ -364,6 +364,27 @@ status_t HwcService::Controls::DisplayModeSetMode(uint32_t display,
   return OK;
 }
 
+status_t HwcService::Controls::EnableHDCPSessionForDisplay(uint32_t display,
+                                                           bool fallback) {
+  mHwc.EnableHDCPSessionForDisplay(display, fallback);
```
[the remark this answers is not in the document]
Comment (kalyankondapally) [reply os/android/hwcservice.cpp]: Handshake may or may not happen during the call. This is just acknowledgement that we have received the request and will act upon in the next frame

**Comment by kalyankondapally** [comment]
@strassek @csravelar ping if you want also to give u feedback.

File: os/android/libhwcservice/test/hwcservice_test.cpp [hunk 5]
```diff
@@ -28,6 +28,10 @@ using namespace hwcomposer;
 static void usage() {
   aout << "Usage: hwcservice_test \n"
           "\t-g: Get current display mode\n"
+          "\t-h: Enable HDCP support for a given Display. \n"
+          "\t-i: Disable HDCP support for a given Display. \n"
+          "\t-j: Enable HDCP support for all displays. \n"
+          "\t-k: Disable HDCP support for all displays. \n"
```
Comment (strassek) [inline os/android/libhwcservice/test/hwcservice_test.cpp]: Drop trailing whitespace

File: os/android/libhwcservice/hwcserviceapi.h [hunk 6]
```diff
@@ -127,23 +127,6 @@ status_t HwcService_Display_SetDeinterlaceParam(HWCSHANDLE hwcs,
...
-
-/// Enumerations for common aspect ratios
-/// Any ratio can be supported, with the upper 16 bits containing one dimension,
-/// the lower 16 bits contains the lower dimension
-typedef enum _EHwcsModeAspectRatio {
-  HWCS_MODE_ASPECT_RATIO_ANY = 0x00000000,
-  HWCS_MODE_ASPECT_RATIO_4_3 = 0x00040003,
-  HWCS_MODE_ASPECT_RATIO_16_9 = 0x00100009,
-} EHwcsModeAspectRatio;
-
```
Comment (strassek) [inline os/android/libhwcservice/hwcserviceapi.h]: Why is this block getting removed? Are you cleaning up stale code? If so, could we put this in a separate patch?

File: common/core/gpudevice.cpp [hunk 7]
```diff
@@ -524,6 +524,43 @@ void GpuDevice::HandleHWCSettings() {
   initialization_state_lock_.unlock();
 }
 
+void GpuDevice::EnableHDCPSessionForDisplay(uint32_t display,
+                                            bool /*fallback*/) {
+  size_t size = total_displays_.size();
```
[the remark this answers is not in the document]
Comment (strassek) [reply common/core/gpudevice.cpp]: Also `size` is only used in 1 place. Could we rewrite the conditional as `if (display > total_displays_.size()) {`?

File: os/android/libhwcservice/hwcserviceapi.cpp [hunk 8]
```diff
@@ -221,6 +221,46 @@ status_t HwcService_DisplayMode_SetMode(HWCSHANDLE hwcs, uint32_t display,
   return pContext->mControls->DisplayModeSetMode(display, config);
 }
 
+status_t HwcService_Video_EnableHDCPSession_ForDisplay(HWCSHANDLE hwcs,
+                                                       uint32_t display,
+                                                       bool fallback) {
+  HwcsContext* pContext = static_cast<HwcsContext*>(hwcs);
+  if (!pContext) {
+    return -1;
```
Comment (strassek) [inline os/android/libhwcservice/hwcserviceapi.cpp]: Given that we are using status_t return type, we ought to use a real return code. I'd suggest `BAD_VALUE`.

**Comment by kalyankondapally** [comment]
@ramalingampc2008 k thanks for the info.

File: os/android/libhwcservice/hwcserviceapi.h (hunk 6 above)
Comment (kalyankondapally) [reply os/android/libhwcservice/hwcserviceapi.h]: dropped from here

File: os/android/libhwcservice/hwcserviceapi.cpp (hunk 8 above)
Comment (kalyankondapally) [reply os/android/libhwcservice/hwcserviceapi.cpp]: ackowledeged, however I would like to keep it consistent with other API and do all clean up once. Added task here, if anyone wants to try this out: https://github.com/intel/IA-Hardware-Composer/issues/268

**Comment by kalyankondapally** [comment]
@ramalingampc2008 updated

**Comment by MicheleL-Intel** [comment]
One issue is that, 'BAD_VALUE' is defined under android namespace, whereas 'status_t' is an 'int'.  If the return type is changed to 'BAD_VALUE' for consistency across the board, then we're confining the return values to strictly android type defines, which is contradictory to IA-HW-Composer being OS-agnostic.  

One possible solution will be to change 'status_t' into an 'enum' to include 'BAD_VALUE'.  This will also minimize the number of changes, including limiting changes to function prototype definitions.

**Comment by MicheleL-Intel** [comment]
Just realized this particular change is strictly for android.

**Comment by kalyankondapally** [comment]
@MicheleL-Intel Future this can be extended outside Android. If you plan to fix the errors, you can come up with a generic sol.

**Comment by kalyankondapally** [comment]
@MicheleL-Intel Also using uint8_t appropriately could go into a separate task, so we unify all API at same time.

**Comment by kalyankondapally** [comment]
merged
